### DISCUSSION TITLE
Embedded Celery Workers in CherryPy

### DISCUSSION THREAD

**Discussion by dmlb2000** [celery/celery#7022] [Ideas]
# Checklist

- [x] I have checked the [issues list](https://github.com/celery/celery/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22Category%3A+Documentation%22+)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues in this issue
      (If there are none, check this box anyway).

## Related Issues and Possible Duplicates

I've not found any documentation on embedding Celery Workers into my Python program. Everything so far has been extending workers to add functionality when the workers start or stop. The closest thing I found was the Py.Test `celery/contrib/testing/app.py` code but I couldn't figure out how, the setup I've got is similar.

#### Related Issues

- None

#### Possible Duplicates

- None

# Description

I'm trying to embed Celery into a CherryPy server. I'm not looking for a lot of concurrency. So, a single process multi-threaded application is desirable for long running tasks that won't return in a reasonable time for web applications.

It seems like both frameworks have good extending documentation with very similar usage patterns. So the reverse could be asked as well, how do I embed a CherryPy server into Celery?
 * https://docs.cherrypy.org/en/latest/extend.html
 * https://docs.celeryproject.org/en/stable/userguide/extending.html

Here's the code I have so far.

```
from threading import Thread
from cherrypy.process import plugins
from celery.bin import worker as celery_worker
from .config import get_config
from .tasks import celery_app

def run_celery_worker():
    """Run the main solo worker."""
    worker = celery_worker.worker(app=celery_app)
    options = {
        'broker': get_config().get('celery', 'broker_url'),
        'loglevel': 'INFO',
        'traceback': True,
        'pool': 'eventlet'
    }
    return worker.run(**options)

class CeleryThreadPlugin(plugins.SimplePlugin):
    def start(self):
        self.bus.log('Starting up Celery worker')
        self.celery_thread = Thread(target=run_celery_worker)
        self.celery_thread.start()

    def stop(self):
        self.bus.log('Stopping down Celery worker')
        celery_app.loader.shutdown_worker()
        sleep(10)
        # The Celery worker.run() isn't returning
        self.bus.log('Celery worker is {}'.format(self.celery_thread.isAlive()))
        self.celery_thread.join()
        self.bus.log('Celery worker joined')
```

I'm using `memory:///` for my `broker_url` similar to the Py.Test integration. The problem is shutdown is not clean. The main script runs and brings up both CherryPy and Celery, when I hit Ctrl-C CherryPy handles the shutdown sequence. The `worker.run()` method doesn't seem to return to stop the worker thread.

I've tried various ways to issue the worker shutdown and none of them seem to make `worker.run()` return.

# Suggestions

It would be nice to have some good documentation on how to embed Celery properly in another application. CherryPy does have similar [docs](https://docs.cherrypy.org/en/latest/deploy.html#embedding-into-another-wsgi-framework) and I've followed them to success embedding CherryPy into specialized Python applications.

Some things to consider:
 * Is it okay for Celery workers to be managed from child Python thread (not main)?
 * What would the API be for that usage?

**Comment by dmlb2000** [comment]
@jaraco may have an idea if this is reasonable or if something else should be done.

**Comment by jaraco** [comment]
I took a brief look and don't have much to add. What you are trying to do sounds sane. I don't understand enough of the Celery internals to make a good assessment on that, but I wish you luck.

**Comment by dmlb2000** [comment]
The thought I had last night was if signals were getting in the way. Is the API I'm using in Celery registering signal handlers that CherryPy `quickstart()` is blowing away?
